pymeasure/instruments/instrument.py

```diff
@@ -327,90 +343,75 @@ class SpecificInstrument(Instrument):
-                raise LookupError("Instrument property can not be set.")
-        else:
-            def fset(self,
-                     value,
-                     set_command=set_command,
-                     validator=validator,
-                     values=values,
-                     map_values=map_values,
-                     set_process=set_process,
-                     check_set_errors=check_set_errors,
-                     ):
-
-                value = set_process(validator(value, values))
+            vals = self.values(command_process(get_command), **kwargs)
+            if check_get_errors:
+                self.check_errors()
+            if len(vals) == 1:
+                value = get_process(vals[0])
                 if not map_values:
-                    pass
+                    return value
                 elif isinstance(values, (list, tuple, range)):
-                    value = values.index(value)
+                    return values[int(value)]
                 elif isinstance(values, dict):
-                    value = values[value]
+                    for k, v in values.items():
+                        if v == value:
+                            return k
+                    raise KeyError("Value {} not found in mapped values".format(value))
                 else:
                     raise ValueError(
                         'Values of type `{}` are not allowed '
                         'for Instrument.control'.format(type(values))
                     )
-                self.write(set_command % value)
-                if check_set_errors:
-                    self.check_errors()
+            else:
+                vals = get_process(vals)
+                return vals
+
+        def fset(self,
+                 value,
+                 set_command=set_command,
+                 validator=validator,
+                 values=values,
+                 map_values=map_values,
+                 set_process=set_process,
+                 command_process=command_process,
+                 check_set_errors=check_set_errors,
+                 ):
+
+            if set_command is None:
+                raise LookupError("Instrument property can not be set.")
+
+            value = set_process(validator(value, values))
+            if not map_values:
+                pass
+            elif isinstance(values, (list, tuple, range)):
+                value = values.index(value)
+            elif isinstance(values, dict):
+                value = values[value]
+            else:
+                raise ValueError(
+                    'Values of type `{}` are not allowed '
+                    'for Instrument.control'.format(type(values))
+                )
+            self.write(command_process(set_command) % value)
+            if check_set_errors:
+                self.check_errors()
 
         # Add the specified document string to the getter
         fget.__doc__ = docs
 
-        if dynamic:
```

Comment:
Don't you wanna keep this signal (to users and developers) that a property is dynamically changeable? I liked this separate factory argument, because "explicit is better than implicit", and most properties use a vanilla python @property, only some will use our dynamic codepath, so the potential impact of bugs is much reduced.

Reply:
Ok, I will revert this